From: David Sterba <dsterba@suse.cz>
Patch-mainline: never
References: FATE#306586
Subject: [PATCH] Btrfs: fix DIO locking type

Btrfs locks buffered and unbuffered writes on it's own and does not expect DIO
code to do any locking manipulation. Pass correct locking type down to DIO code
and prevent deadlock (visible via xfstests/133).

This adaptaion to current block layer code, but the codepath taken in
__blockdev_direct_IO matches upstream.

Signed-off-by: David Sterba <dsterba@suse.cz>

--- a/linux-2.6/fs/btrfs/inode.c	2011-06-29 16:32:53.000000000 +0200
+++ b/linux-2.6/fs/btrfs/inode.c	2011-06-30 14:51:25.000000000 +0200
@@ -6211,7 +6211,7 @@
 	ret = __blockdev_direct_IO(rw, iocb, inode,
 		   BTRFS_I(inode)->root->fs_info->fs_devices->latest_bdev,
 		   iov, offset, nr_segs, btrfs_get_blocks_direct, NULL,
-		   btrfs_submit_direct, 0);
+		   btrfs_submit_direct, DIO_NO_LOCKING);
 
 	if (ret < 0 && ret != -EIOCBQUEUED) {
 		clear_extent_bit(&BTRFS_I(inode)->io_tree, offset,
